### PR TITLE
[TRUS-586] - What is [trustee name]'s address?

### PR DESCRIPTION
**Pull request by jassalrichy**
[TRUS-586] - What is [trustee name]'s address?

### CODE REVIEW

File: app/utils/CheckYourAnswersHelper.scala [hunk 1]
```diff
@@ -30,6 +30,18 @@ import viewmodels.AnswerRow
 
 class CheckYourAnswersHelper @Inject()(countryOptions: CountryOptions)(userAnswers: UserAnswers)(implicit messages: Messages) {
 
+  def trusteesNino(index: Int): Option[AnswerRow] = userAnswers.get(TrusteesNinoPage(index)) map {
+    x =>
+      AnswerRow(
+        "trusteesNino.checkYourAnswersLabel",
+        HtmlFormat.escape(x),
+        routes.TrusteesNinoController.onPageLoad(CheckMode, index).url,
+        getTrusteeFirstLastName(index)
```
Comment (adamconder) [inline app/utils/CheckYourAnswersHelper.scala]: This was refactored into the companion object. Can you please call `trusteeName(index, userAnswers)` instead.

File: app/utils/CheckYourAnswersHelper.scala [hunk 2]
```diff
@@ -30,6 +30,18 @@ import viewmodels.AnswerRow
...
+      )
+  }
+
+  def getTrusteeFirstLastName(index: Int): String = userAnswers.get(TrusteesNamePage(index)).get.toString
```
Comment (adamconder) [inline app/utils/CheckYourAnswersHelper.scala]: Please remove this method as it duplicates what is in the companion object.

File: test/forms/TrusteesUkAddressFormProviderSpec.scala [hunk 3]
```diff
@@ -75,5 +125,38 @@ class TrusteesUkAddressFormProviderSpec extends StringFieldBehaviours {
...
+    )
+
+  }
+
+  ".postcode" must {
+
+    val fieldName = "postcode"
+    val requiredKey = "trusteesUkAddress.error.postcode.required"
+
+    behave like fieldThatBindsValidData(
```
Comment (adamconder) [inline test/forms/TrusteesUkAddressFormProviderSpec.scala]: We have tested this differently elsewhere for postcodes.

```scala
behave like fieldWithRegexpWithGenerator(
      form,
      fieldName,
      regexp = postcodeRegexWithSpaceConstrained,
      generator = arbitrary[String],
      error = FormError(fieldName, invalidKey)
    )
```

File: test/views/TrusteesUkAddressViewSpec.scala [hunk 4]
```diff
@@ -51,7 +51,11 @@ class TrusteesUkAddressViewSpec extends QuestionViewBehaviours[TrusteesUkAddress
       applyView,
       messageKeyPrefix,
       routes.TrusteesUkAddressController.onSubmit(NormalMode, index).url,
-      "line1", "line2", "line3", "townOrCity", "postcode"
+      Seq("line1", "line2", "line3", "townOrCity", "postcode"),
+      trusteeName
     )
```
Comment (adamconder) [inline test/views/TrusteesUkAddressViewSpec.scala]: Do we need to pass in hint text tto assert?

File: app/controllers/TrusteesUkAddressController.scala [hunk 5]
```diff
@@ -77,8 +77,10 @@ class TrusteesUkAddressController @Inject()(
           Future.successful(BadRequest(view(formWithErrors, mode, index, trusteeName))),
 
         value => {
+          val transformedPostcode = value.postcode.toUpperCase()
```
Comment (adamconder) [inline app/controllers/TrusteesUkAddressController.scala]: Can we do this as part of the Mapping `postcode()` type? We already trim the postcode there.

File: test/navigation/navigators/TrusteeRoutes.scala [hunk 6]
```diff
@@ -173,11 +173,20 @@ trait TrusteeRoutes {
         }
       }
 
-      "go to TrusteeTelephoneNumberPage from TrusteeLivesInUKPage" in {
+      "go to TrusteesUkAddressPage from TrusteeLivesInUKPage" in {
         forAll(arbitrary[UserAnswers]) {
           userAnswers =>
```
Comment (adamconder) [inline test/navigation/navigators/TrusteeRoutes.scala]: This should be constrained to only navigate if they have answered 'Yes' to living in the UK.

File: app/models/TrusteesUkAddress.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                                       line4: Option[String],
+                                       country: String
+                                     ) extends Address
+
+object InternationalAddress {
+
+  implicit lazy val formats: OFormat[InternationalAddress] = Json.format[InternationalAddress]
+}
+
+trait Address
```
Comment (adamconder) [inline app/models/TrusteesUkAddress.scala]: Can we make this a sealed trait.

File: app/utils/CheckYourAnswersHelper.scala [hunk 8]
```diff
@@ -50,6 +50,15 @@ class CheckYourAnswersHelper @Inject()(countryOptions: CountryOptions)(userAnswe
       )
   }
 
+  def trusteesUkAddress(index: Int): Option[AnswerRow] = userAnswers.get(TrusteesUkAddressPage(index)) map {
+    x =>
+      AnswerRow(
+        "trusteesUkAddress.checkYourAnswersLabel",
+        ukAddress(x),
+        routes.TrusteesUkAddressController.onPageLoad(CheckMode, index).url
+      )
```
Comment (adamconder) [inline app/utils/CheckYourAnswersHelper.scala]: Do you need to pass through trusteeName as an argument?

File: app/views/TrusteesUkAddressView.scala.html [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    title = s"${errorPrefix(form)} ${messages("trusteesUkAddress.title", trusteeName)}"
+    ) {
+
+    @formHelper(action = TrusteesUkAddressController.onSubmit(mode, index), 'autoComplete -> "off") {
+
+        @components.back_link()
+
+        @components.error_summary(form.errors)
+
+        @components.heading("trusteesUkAddress.heading", optionalParam = Some(trusteeName))
```
Comment (adamconder) [inline app/views/TrusteesUkAddressView.scala.html]: Does this need to be Some(trusteeName)? Can't we just pass trusteeName? I may not be understanding a change we have done here.

File: app/models/TrusteesUkAddress.scala (hunk 7 above)
Comment (jassalrichy) [reply app/models/TrusteesUkAddress.scala]: Done

File: app/utils/CheckYourAnswersHelper.scala (hunk 8 above)
Comment (jassalrichy) [reply app/utils/CheckYourAnswersHelper.scala]: Done

File: app/views/TrusteesUkAddressView.scala.html (hunk 9 above)
Comment (jassalrichy) [reply app/views/TrusteesUkAddressView.scala.html]: optionalParam iof type Option[String] within component heading. If not provided then not passed into the message within heading component:

@(headingKey: String, optionalParam: Option[String] = None, headingSize: String = "heading-large")(implicit messages: Messages)
<h1 class="@headingSize">@messages(headingKey, if(optionalParam.isDefined) {optionalParam.get} else "")</h1>

File: test/forms/TrusteesUkAddressFormProviderSpec.scala (hunk 3 above)
Comment (jassalrichy) [reply test/forms/TrusteesUkAddressFormProviderSpec.scala]: Refactored to:

    behave like fieldWithRegexpWithGenerator(
      form,
      fieldName,
      regexp = Validation.postcodeRegex,
      generator = arbitrary[String],
      error = FormError(fieldName, invalidKey, Seq(Validation.postcodeRegex))
    )

File: test/navigation/navigators/TrusteeRoutes.scala (hunk 6 above)
Comment (jassalrichy) [reply test/navigation/navigators/TrusteeRoutes.scala]: Done.

File: test/views/TrusteesUkAddressViewSpec.scala (hunk 4 above)
Comment (jassalrichy) [reply test/views/TrusteesUkAddressViewSpec.scala]: Done. Refactor of QuestionViewBehaviours.pageWithTextFields to accept an optional hit text for pageWithTextFields in order to asset

File: app/controllers/TrusteesUkAddressController.scala (hunk 5 above)
Comment (jassalrichy) [reply app/controllers/TrusteesUkAddressController.scala]: Done.

File: test/forms/TrusteesUkAddressFormProviderSpec.scala (hunk 3 above)
Comment (jassalrichy) [reply test/forms/TrusteesUkAddressFormProviderSpec.scala]: Done

File: app/utils/CheckYourAnswersHelper.scala (hunk 1 above)
Comment (jassalrichy) [reply app/utils/CheckYourAnswersHelper.scala]: Done

File: app/utils/CheckYourAnswersHelper.scala (hunk 2 above)
Comment (jassalrichy) [reply app/utils/CheckYourAnswersHelper.scala]: Done

**Pull request merged**